### PR TITLE
Add resampling offset

### LINKED ISSUE #1711: Equivalent `offset` argument to Pandas resample

### PR DESCRIPTION
**Pull request by vasil-pashov**
#### Reference Issues/PRs

Resolve #1711
#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/version/python_bindings.cpp [hunk 1]
```diff
@@ -27,6 +27,46 @@
...
+    timestamp start,
+    timestamp end,
+    std::string_view rule,
+    ResampleBoundary closed_boundary_arg,
+    timestamp offset
+) {
+    const timestamp rule_ns = [&rule]() {
+        py::gil_scoped_acquire acquire_gil;
+        return python_util::pd_to_offset(rule);
+    }();
```
Comment (alexowens90) [inline cpp/arcticdb/version/python_bindings.cpp:59]: Adding a lambda just to add a scope seems less clean to me than
```
timestamp rule_ns;
{
    py::gil_scoped_acquire acquire_gil;
    rule_ns = python_util::pd_to_offset(rule);
}
```

File: python/tests/unit/arcticdb/version_store/test_resample.py [hunk 2]
```diff
@@ -503,4 +505,94 @@ def test_resampling_empty_type_column(lmdb_version_store_empty_types_v1):
...
+        lib.read(sym, query_builder=q)
+
+class TestResamplingOffset:
+
+    @staticmethod
+    def all_aggregations_dict(col):
+        return {f"to_{agg}": (col, agg) for agg in ALL_AGGREGATIONS}
+
+    @pytest.mark.parametrize("closed", ["left", "right"])
+    def test_offset_smaller_than_freq(self, lmdb_library, closed):
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_resample.py:517]: Why not `lmdb_version_store_v1`? (for all test cases)

File: python/tests/unit/arcticdb/version_store/test_resample.py [hunk 3]
```diff
@@ -503,4 +505,94 @@ def test_resampling_empty_type_column(lmdb_version_store_empty_types_v1):
...
\ No newline at end of file
+        lib.read(sym, query_builder=q)
+
+class TestResamplingOffset:
+
+    @staticmethod
+    def all_aggregations_dict(col):
+        return {f"to_{agg}": (col, agg) for agg in ALL_AGGREGATIONS}
+
+    @pytest.mark.parametrize("closed", ["left", "right"])
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_resample.py:516]: As all tests have the same parametrization, you can directly parametrize `TestResamplingOffset`

File: python/tests/unit/arcticdb/version_store/test_resample.py [hunk 4]
```diff
@@ -503,4 +505,94 @@ def test_resampling_empty_type_column(lmdb_version_store_empty_types_v1):
     q = QueryBuilder()
     q = q.resample("s").agg({"col": "first"})
     with pytest.raises(SchemaException):
-        lib.read(sym, query_builder=q)
\ No newline at end of file
+        lib.read(sym, query_builder=q)
+
+class TestResamplingOffset:
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_resample.py:510]: We should also test using `pd.Timedelta` for the offset argument

**Review by alexowens90 — CHANGES_REQUESTED** [review]

File: python/tests/unit/arcticdb/version_store/test_resample.py (hunk 2 above)
Comment (vasil-pashov) [reply python/tests/unit/arcticdb/version_store/test_resample.py:517]: No particular reason, `lmdb_version_store_v1` is library V1 `lmdb_library` is V2. Should I change it to `lmdb_version_store_v1` to match the other tests?

File: cpp/arcticdb/version/python_bindings.cpp (hunk 1 above)
Comment (vasil-pashov) [reply cpp/arcticdb/version/python_bindings.cpp:59]: That's trick from Herb Sutter for initialization of constant complex variables from Herb Sutter: https://herbsutter.com/2013/04/05/complex-initialization-for-a-const-variable/ I don't mind changing it if you insist.

File: python/tests/unit/arcticdb/version_store/test_resample.py (hunk 4 above)
Comment (vasil-pashov) [reply python/tests/unit/arcticdb/version_store/test_resample.py:510]: Cool, I didn't know you can parametrize the whole suite.

File: python/tests/unit/arcticdb/version_store/test_resample.py (hunk 3 above)
Comment (vasil-pashov) [reply python/tests/unit/arcticdb/version_store/test_resample.py:516]: Cool, I didn't know you can parametrize the whole suite.

File: python/tests/unit/arcticdb/version_store/test_resample.py (hunk 2 above)
Comment (alexowens90) [reply python/tests/unit/arcticdb/version_store/test_resample.py]: Yeh, unless there is a good reason not to, `lmdb_version_store_v1 ` should be the default

File: cpp/arcticdb/version/python_bindings.cpp (hunk 1 above)
Comment (alexowens90) [reply cpp/arcticdb/version/python_bindings.cpp]: Yeh keeping it `const` is nice, happy to keep it as a lambda

**Review by alexowens90 — APPROVED** [review]

**Pull request merged**